### ISSUE TITLE
memo in createComponent

### ISSUE THREAD

**Issue by titoBouzout** [ryansolid/mobx-jsx#21]
On the last line of this function it seems to be creating a memo in some situation. Although I tried to print it I had no luck using class components and even function components. So Im confused, that code runs at all?.  I would like to have control about whats memoized and not memoized for in case of /render once/ . Whats the purpose of the memoizing? can it be removed please if the only purpose was to just memoize everything for the proof of concept. Thanks! 

```javascript 

function createComponent(Comp, props, dynamicKeys) {
  console.log('creating compooent', Comp)
  if (dynamicKeys) {
    for (let i = 0; i < dynamicKeys.length; i++) dynamicProperty(props, dynamicKeys[i])
  }

  let c

  if (Comp.prototype && Comp.prototype.isClassComponent) {
    console.log('is class')
    c = (0, _mobx.untracked)(() => {
      const comp = new Comp(props)
      return comp.render(props)
    })
  } else {
    console.log('is not class')
    c = (0, _mobx.untracked)(() => Comp(props))
  }
  console.log('component is', c)
  return typeof c === 'function' ? memo(c) : c
}

```

**Comment by ryansolid** [comment]
Yeah I can change this. See if a component returns a function I was having it auto memoize. Most components don't return functions. Only ones that output needs to be dynamic at the root. Otherwise this should just get passed over. The necessity of this comes in when you have arrays with multiple inserts each with their own dynamic parts. You don't want to trigger re-evaluation in the other parts. Stuff that renders once is fine since it has already evaluated by the time the insert happens but dynamic parts would need to. The reason being is that if we are returning functions we don't know the value of that function unless we execute it. And without knowing the cost of the evaluation it was simpler to just make it a reactive memo to prevent extra work. 

I was doing this with the compiler for a bit but it was cause random issues with reactive control flow in arrays. Basically things that managed arrays and inserted in a way the compiler couldn't see from static analysis. In Solid I have moved back to this being explicit in control flow. But I left this here since for the most part you need to do this memoization anyway. Of course the downside is this approach can pay extra cost with many layers of nesting. Like if a Component returns a Component that returns a dynamic function (memo) we end up wrapping multiple times. So I will change this back to being explicit in these libraries as well.

That being said more than likely this code is probably never running for your components.

**Comment by titoBouzout** [comment]
Thanks for the explanation, I understand. Now Im not completely sure I want it removed as I maybe dont understand in the near future this edge case, if you remove it can you please add examples on when this is needed.

I was just able to trigger a memoization by using lazy,  gave me: `() => rendered.get()`

**Comment by ryansolid** [comment]
Yeah the consideration is when the consequence of re-evaluation is heavy. Like just forwarding an boxed observable is completely unnecessary for the extra wrapper. It is more for things like lists and conditionals. Since I auto wrap ternaries and boolean expressions in the JSX and I provide a map helper isn't really necessary. As long as people use those there is no concern. It was just a way to keep things predictable at a fairly cheap cost most of the time. I went back to explicit with Solid and probably should here. Only the last minor version or so had this change.

What I mean is picture:
```js
() => this.show ? doExpensive() : doOtherExpensive()
```

You'd never want this to execute more than it needs to. My "memo" is just a "computed" in MobX terms more or less with recursive value evaluation and conditional notification. I just implemented it with signal and autorun since it was simpler given the way the code was. There might be a way to just do it with a computed that way.

**Comment by ryansolid** [comment]
While it's possible to hit this scenario, unlike Solid which has control flow components, you are much more unlikely to. I intend to move more and more of this logic into compiler time, and since there are no cases today with shipped things that require this extra care and it's an extra overhead I'm going to put things back to how they've always been before the latest versions of the library. Largely a lot of stuff becomes copy and paste from Solid and I didn't give the specific library enough consideration.

Note to self:
I need to still find the best balance. The challenge being only the insertion point needs to be wrapped which I can do mostly automatically at compile time with the exception of things that produce arrays with dynamic elements programmatically instead of through JSX. But since the libraries controls those array helpers I can modify them instead of pushing it on to end user who doesn't know where their component will be used. In one sense the end user is the best to know if their creation is expensive and needs wrapping but that can lead to unnecessary levels of nested wrapping. But it is fairly safe to say any case where an array is constructed regardless of the lookup cost if direct indexes are dynamic we should be memorizing to prevent unnecessary re-evaluation between indexes. And only very specific transformations that involve accessing resolved values multiple times would need memorization in the middle.

**Comment by ryansolid** [comment]
Released in 0.11.7
